cpp/bench/ann/src/common/dataset.hpp

```diff
@@ -19,526 +19,916 @@
+          size_t size = data_end - data_start;
+          mmap_owner owner{size, flags};
+          std::fseek(file_.descriptor().value(), data_start, SEEK_SET);
+          size_t n_elems = file_.rows_limit() * file_.n_cols();
+          if (std::fread(owner.data(), sizeof(T), n_elems, file_.descriptor().value()) != n_elems) {
+            throw std::runtime_error{"cuvs::bench::blob_mmap() fread " + file_.path() + " failed"};
+          }
+          handle_.emplace(std::move(owner), 0);
+        } else {
+          // Map the file
+          // If this is a temporary file, we're supposed to write to it, hence MAP_SHARED.
+          flags |= file_.is_temporary() ? MAP_SHARED : MAP_PRIVATE;
+          size_t mmap_start = (data_start / page_size) * page_size;
+          size_t mmap_size  = data_end - mmap_start;
+          handle_.emplace(
+            mmap_owner{file_.descriptor(), mmap_start, mmap_size, flags, file_.is_temporary()},
+            data_start - mmap_start);
+        }
+      } catch (const mmap_error& e) {
+        bool hugepages_2mb_asked = hugepages_2mb_requested_ == HugePages::kAsk ||
+                                   hugepages_2mb_requested_ == HugePages::kRequire;
+        if (e.code() == EPERM && hugepages_2mb_asked && hugepages_2mb_actual_) {
+          if (hugepages_2mb_requested_ == HugePages::kRequire) {
+            log_warn(
+              "cuvs::bench::blob_mmap: `mmap` failed to map due to EPERM, which is likely caused "
+              "by the permissions issue. You either need a CAP_IPC_LOCK capability or run the "
+              "program with sudo. We will try again without huge pages.");
+          }
+          hugepages_2mb_actual_ = false;
+          return handle();
+        }
+        if (e.code() == EINVAL && hugepages_2mb_asked && hugepages_2mb_actual_ &&
+            !copy_in_memory_) {
+          if (hugepages_2mb_requested_ == HugePages::kRequire) {
+            log_warn(
+              "cuvs::bench::blob_mmap: `mmap` failed to map due to EINVAL, which is likely caused "
+              "by the file system not supporting huge pages. We will try again without huge "
+              "pages.");
+          }
+          hugepages_2mb_actual_ = false;
+          return handle();
+        }
+        throw;  // The error is not due to huge pages or otherwise unrecoverable
       }
-      nrows_ -= subset_first_row_;
     }
-    if (subset_size_) { nrows_ = subset_size_; }
+    return handle_.value();
   }
-}
 
-template <typename T>
-void bin_file<T>::check_suffix()
-{
-  auto pos = file_.rfind('.');
-  if (pos == std::string::npos) {
-    throw std::runtime_error{"name of bin_file doesn't have a suffix: " + file_};
+ public:
+  [[nodiscard]] auto data() const -> T*
+  {
+    auto& [owner, offset] = handle();
+    return reinterpret_cast<T*>(reinterpret_cast<uint8_t*>(owner.data()) + offset);
   }
-  std::string suffix = file_.substr(pos + 1);
 
-  if constexpr (std::is_same_v<T, float>) {
```

Comment:
It is great to see these test gone. File types were already checked in advance by config, so these tests were not useful, and were sometimes very inconvenient.